### PR TITLE
DO-1320 Update RATU converter

### PR DESCRIPTION
**Pull request by Hedgehog18**
The file from February 9 was flooded into the database, after that the file from March 2 was transferred to the converter.
After working out the converter:
District - nothing deleted
City - 616 records tagged "deleted_at"
City District - 393 records tagged "deleted_at"
Street - 1395 records tagged "deleted_at"
Converter run time is approximately 7 minutes.

### CODE REVIEW

**Comment by roman-tiukh** [comment]
@Iva-khar @IegorSoboliev review pls

File: location_register/converter/ratu.py [hunk 1]
```diff
@@ -66,6 +74,9 @@ def save_or_get_district(self, name, region):
             )
             self.all_districts_dict[district_code] = district
             return district
+        else:
+            if self.outdated_districts_dict.get(district_code):
```
Comment (Iva-khar) [inline location_register/converter/ratu.py:78]: It's redudance checking

File: location_register/converter/ratu.py [hunk 2]
```diff
@@ -81,6 +92,9 @@ def save_or_get_city(self, name, region, district):
             )
             self.all_cities_dict[city_code] = city
             return city
+        else:
+            if self.outdated_cities_dict.get(city_code):
```
Comment (Iva-khar) [inline location_register/converter/ratu.py:96]: the same

File: location_register/converter/ratu.py [hunk 3]
```diff
@@ -113,6 +130,9 @@ def save_street(self, name, region, district, city, citydistrict):
                 name=street_name,
                 code=street_code
             )
+        else:
+            if self.outdated_streets_dict.get(street_code):
```
Comment (Iva-khar) [inline location_register/converter/ratu.py:134]: the same

File: location_register/converter/ratu.py [hunk 4]
```diff
@@ -96,6 +110,9 @@ def save_or_get_citydistrict(self, name, region, district, city):
             )
             self.all_citydistricts_dict[citydistrict_code] = citydistrict
             return citydistrict
+        else:
+            if self.outdated_citydistricts_dict.get(citydistrict_code):
```
Comment (Iva-khar) [inline location_register/converter/ratu.py:114]: the same

File: location_register/converter/ratu.py [hunk 5]
```diff
@@ -113,6 +130,9 @@ def save_street(self, name, region, district, city, citydistrict):
                 name=street_name,
```
Comment (Iva-khar) [inline location_register/converter/ratu.py:130]: What about the same street names in different some nemes cities? line 123

File: location_register/converter/ratu.py [hunk 6]
```diff
@@ -96,6 +110,9 @@ def save_or_get_citydistrict(self, name, region, district, city):
             )
```
Comment (Iva-khar) [inline location_register/converter/ratu.py:110]: line 102 `citydistrict_code = city.name + citydistrict_name` is it unic parametr?

**Review by Iva-khar — CHANGES_REQUESTED** [review]

File: location_register/converter/ratu.py (hunk 1 above)
Comment (Hedgehog18) [reply location_register/converter/ratu.py:78]: Without this check there is an error, yesterday with Egor this moment was discussed

File: location_register/converter/ratu.py (hunk 3 above)
Comment (Hedgehog18) [reply location_register/converter/ratu.py:134]: Without this check there is an error, yesterday with Egor this moment was discussed

File: location_register/converter/ratu.py (hunk 4 above)
Comment (Hedgehog18) [reply location_register/converter/ratu.py:114]: Without this check there is an error, yesterday with Egor this moment was discussed

File: location_register/converter/ratu.py (hunk 5 above)
Comment (Hedgehog18) [reply location_register/converter/ratu.py:130]: We refer to the "code": street_code = city.name + street_name.
If I understood the question correctly.

File: location_register/converter/ratu.py [hunk 7]
```diff
@@ -128,6 +148,20 @@ def save_to_db(self, records):
...
+        if len(self.outdated_cities_dict):
+            for cities in self.outdated_cities_dict.values():
+                cities.soft_delete()
+        if len(self.outdated_citydistricts_dict):
+            for citydistricts in self.outdated_citydistricts_dict.values():
+                citydistricts.soft_delete()
+        if len(self.outdated_streets_dict):
+            for streets in self.outdated_streets_dict.values():
+                streets.soft_delete()
+
```
Comment (IegorSoboliev) [inline location_register/converter/ratu.py:164]: Actually, we have no reason to measure a length here

File: location_register/converter/ratu.py (hunk 4 above)
Comment (Iva-khar) [reply location_register/converter/ratu.py:114]: What error? Why error?

Comment (roman-tiukh) [reply location_register/converter/ratu.py:114]:
```
d = {'hello': 1}
del d['as']
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
KeyError: **'as'**
```

Comment (roman-tiukh) [reply location_register/converter/ratu.py:114]:
```
d = {'hello': 1}
del d['as']
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
KeyError: **'as'**
```

**Review by roman-tiukh — APPROVED** [review]

File: location_register/converter/ratu.py (hunk 5 above)
Comment (Iva-khar) [reply location_register/converter/ratu.py:130]: Not. It have to be unic.

File: location_register/converter/ratu.py (hunk 6 above)
Comment (Iva-khar) [reply location_register/converter/ratu.py:110]: Must be `citydistrict_code = region.name + city.name + citydistrict_name` for sure

File: location_register/converter/ratu.py (hunk 5 above)
Comment (Iva-khar) [reply location_register/converter/ratu.py:130]: Must be `street_code = region.name + district.name + city.name + citydistrict_name` for unic

**Review by Iva-khar — CHANGES_REQUESTED** [review]

**Comment by Hedgehog18** [comment]
After testing the updated converter, the results changed:
CityDistrict: 519 "deleted" (old result - 393)
Street: 1636 "deleted" (old result - 1395)

File: location_register/converter/ratu.py [hunk 8]
```diff
@@ -96,14 +110,18 @@ def save_or_get_citydistrict(self, name, region, district, city):
...
         return self.all_citydistricts_dict[citydistrict_code]
 
     def save_street(self, name, region, district, city, citydistrict):
         street_name = change_to_full_name(name)
         # Saving streets that are located in Kyiv and Sevastopol that are regions
         if not city:
             city = self.save_or_get_city(region.name, region, district)
-        street_code = city.name + street_name
+        district_name = 'EMPTY' if not district else district.name
+        street_code = region.name + district_name + city.name + street_name
```
Comment (Iva-khar) [inline location_register/converter/ratu.py:124]: I would add also sitydistrict.name here

File: location_register/converter/ratu.py (hunk 1 above)
Comment (Iva-khar) [reply location_register/converter/ratu.py:78]: I return to my mean that it's redudance checking. When you have unic field 'code' error with missing key never happens. You have wrong 'code' if it happens

File: location_register/converter/ratu.py (hunk 2 above)
Comment (Iva-khar) [reply location_register/converter/ratu.py:96]: the same

File: location_register/converter/ratu.py (hunk 4 above)
Comment (Iva-khar) [reply location_register/converter/ratu.py:114]: I return to my mean that it's redudance checking. When you have unic field 'code' error with missing key never happens. You have wrong 'code' if it happens

File: location_register/converter/ratu.py (hunk 3 above)
Comment (Iva-khar) [reply location_register/converter/ratu.py:135]: the same

**Review by Iva-khar — CHANGES_REQUESTED** [review]

File: location_register/converter/ratu.py (hunk 3 above)
Comment (Iva-khar) [reply location_register/converter/ratu.py:136]: test without this cheking

**Review by Iva-khar — CHANGES_REQUESTED** [review]

**Review by Iva-khar — APPROVED** [review]

**Pull request merged**